logstash-core/lib/logstash/agent_plugin_manager.rb

```diff
@@ -0,0 +1,65 @@
+# encoding: utf-8
+require "logstash/agent"
+
+module LogStash module AgentPluginManager
+  DEFAULT_AGENT_NAME = :default
+  class DuplicatePluginNameError < Error; end
+
+  REGISTRY = {}
+
+  # Reset plugin registry to just the default plugin
+  def self.reset!
+    REGISTRY.clear
+    REGISTRY[DEFAULT_AGENT_NAME] = LogStash::Agent
+  end
+
+  reset!
+
+  # Search gems for available plugins and load their libs
+  def self.load_all
+    find_plugins.each do |plugin|
+      name = plugin.name.split('-')[-1]
+      require "logstash/agents/#{name}"
+    end
+  end
+
+  # Return a list of Gem::Specification s that start with logstash-agent-
+  def self.find_plugins
+    Gem::Specification.find_all{|spec| spec.name =~ /logstash-agent-/ }
+  end
+
+  # To be called by a plugin when its class is first loaded
+  # Plugins should call this with the following code:
+  #
+  # require 'lib/logstash/agent_plugin_manager'
+  #
+  # class MyLogStashAgent < LogStash::Agent
+  #   LogStash::AgentPluginManager.register(:my_logstash_agent, self)
+  #
+  #   # ...
+  # end
+  def self.register(name, plugin_class)
+    name_sym = name.to_sym
+
+    if (conflicting_class = registry[name_sym])
```

Comment:
nope, so this is tied not to "laws" but to personal feelings. I speak by myself, this kind of constructors are confusing and error prone to me. I would recommend rewriting, but your call at the end of the day.